osbuild/mounts.py

```diff
@@ -58,7 +58,21 @@ def __init__(self, devices: DeviceManager, root: str) -> None:
 
     def mount(self, mount: Mount) -> Dict:
 
-        source = self.devices.device_abspath(mount.device)
+        # Get the absolute path to the source device inside the
+        # temporary filesystem (i.e. /run/osbuild/osbuild-dev-xyz/loop0)
+        # and also the relative path to the source device inside
+        # that filesystem (i.e. loop0). If the device also exists on the
+        # host in `/dev` (like /dev/loop0), we'll use that path for the
+        # mount because some tools (like grub2-install) consult mountinfo
+        # to try to canonicalize paths for mounts and inside the bwrap env
+        # the device will be under `/dev`. https://github.com/osbuild/osbuild/issues/1492
+        source = None
+        relpath = self.devices.device_relpath(mount.device)
+        abspath = self.devices.device_abspath(mount.device)
+        if relpath and os.path.exists(os.path.join('/dev', relpath)):
+            source = os.path.join('/dev', relpath)
+        elif abspath:
```

Comment:
This could leave source set to None, which I'm guessing will result in a cascade of failures. I'd just fall back to the previous behavior if the relpath doesn't exist.

Reply:
the behavior is already such that source could be set to `None`:

https://github.com/osbuild/osbuild/blob/05c0fd319de2767047310d750a1e0c4e9a8f8cc8/osbuild/devices.py#L63-L72


I guess what you want is just this?

```
else:
    source = abspath
```

I can make that change but I really don't think it's functionally different than what I have.

Reply:
I honestly didn't realize `device_abspath` could return `None` either until `mypy` was giving me warnings like:

```
osbuild/mounts.py:68: error: Argument 2 to "join" has incompatible type "Optional[str]"; expected "str"
```

Reply:
Yeah, it seems easier to understand if it's just an else, or I'm more inclined to do things like:
```
source = self.devices.device_abspath(mount.device)
if relpath and os.path.exists(os.path.join('/dev', relpath)):
    source = os.path.join('/dev', relpath)
```

But that's just a style preference.

Reply:
ok I implemented this